src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala

```diff
@@ -122,6 +122,13 @@ case class WriteIntoDelta(
       case _ => Nil
     }
 
-    newFiles ++ deletedFiles
+    val noDataChange = options.noDataChange
+    if (noDataChange) {
+      newFiles ++ deletedFiles
+    }
+    else {
+      newFiles.map(_.copy(dataChange = noDataChange)) ++
+        deletedFiles.map(_.copy(dataChange = noDataChange))
+    }
```

Comment:
I think these cases are backwards. You want the files to have dataChange = false if there's no data change. 

Other option is to rename noDataChange above to rearrangeOnly if that's less confusing.